src/Control/Email/Email.php

```diff
@@ -753,215 +370,145 @@ public function getHTMLTemplate()
-                $this->getSwiftMessage()->addPart($plainPart, 'text/plain', 'utf-8');
-            }
-        } else {
-            if ($plainPart) {
-                $this->setBody($plainPart);
-            }
-            $this->getSwiftMessage()->setContentType('text/plain');
-            $this->getSwiftMessage()->setCharset('utf-8');
+        // Plain render fails over to generated from html
+        if (!$plainRender && $htmlRender) {
+            // call html_entity_decode() to ensure any encoded HTML is also stripped inside ->Plain()
+            $dbField = DBField::create_field('HTMLFragment', html_entity_decode($htmlRender));
+            $plainRender = $dbField->Plain();
         }
 
-        return $this;
-    }
+        // Remove the $htmlRender if was only used as fail over to create $plainRender
+        if ($plainOnly && $htmlRender) {
+            $htmlRender = null;
+        }
 
-    /**
-     * @return Swift_MimePart|false
-     */
-    public function findPlainPart()
-    {
-        foreach ($this->getSwiftMessage()->getChildren() as $child) {
-            if ($child instanceof Swift_MimePart && $child->getContentType() == 'text/plain') {
-                return $child;
-            }
+        // Handle edge case where data changed since last render and there's no template
+        if (!$htmlRender && $htmlBody) {
+            $htmlRender = $htmlBody;
         }
-        return false;
-    }
 
-    /**
-     * @return bool
-     */
-    public function hasPlainPart()
-    {
-        if ($this->getSwiftMessage()->getContentType() === 'text/plain') {
-            return true;
+        if (!$plainRender && $plainBody) {
+            $plainRender = $plainBody;
         }
-        return (bool) $this->findPlainPart();
-    }
 
-    /**
-     * Automatically adds a plain part to the email generated from the current Body
-     *
-     * @return $this
-     */
-    public function generatePlainPartFromBody()
-    {
-        $plainPart = $this->findPlainPart();
-        if ($plainPart) {
-            $this->getSwiftMessage()->detach($plainPart);
+        // Rendering is finished
+        Requirements::restore();
+
+        if ($plainRender) {
+            $this->text($plainRender);
+        }
+        if ($htmlRender) {
+            $this->html($htmlRender);
         }
-        unset($plainPart);
 
-        $this->getSwiftMessage()->addPart(
-            Convert::xml2raw($this->getBody()),
-            'text/plain',
-            'utf-8'
-        );
+        $this->dataAtLastRender = $this->data;
 
         return $this;
     }
+
+    private function getDataHasChangedSinceLastRender(): bool
+    {
+        if (is_array($this->data) !== is_array($this->dataAtLastRender)) {
+            return true;
+        }
+        if (is_array($this->data)) {
+            return $this->data !== $this->dataAtLastRender;
+        }
+        // ViewableData
+        return serialize($this->data) !== serialize($this->dataAtLastRender);
```

Comment:
We should use `getData()` here - extensions or subclasses might dynamically change the data.

Reply:
Updated